### PR TITLE
Bratseth/output immediate graph

### PR DESCRIPTION
**Pull request by bratseth**
@lesters

This is some work I did for https://github.com/vespa-engine/vespa/issues/9662

- Pretty printing of the intermediate graph in logs and some error messages. WIP as I haven't added output for all the nodes yet.
- Logging of the constraint problem and solution, and in some error messages.
- Allow ExpandDim to expand after the last dimension (this was an outright bug).
- Make soft constraints soft.
- Add inequality constraints for dimensions injected from constants and arguments.

I think the last item necessitates a tree-based constraint solving algorithm as currently we're unable to find a solution. Or maybe there are some shortcuts?

I'm at the end of the day here now. If you have some ideas on how to do that feel free to add to this  while I'm resting.

### CODE REVIEW

**Comment by bratseth** [comment]
On closer examination it seems like these constraints actually cannot be resolved:

```
  input_embedding_user_guid_0 != input_embedding_user_guid_1
  input_embedding_user_guid_0 < test_dense_out_kernel_1 
  Dot_ExpandDims_1 == test_batch_normalization_v1_moving_variance_0 
  test_dense_out_kernel_0 != test_dense_out_kernel_1
  test_dense_out_kernel_1 == test_batch_normalization_v1_moving_mean_0
  test_batch_normalization_v1_moving_mean_0 == test_batch_normalization_v1_moving_variance_0 
  Dot_ExpandDims_1 == test_dense_out_kernel_0
  Dot_ExpandDims_1 == test_batch_normalization_v1_moving_mean_0
```

This requires that test_dense_out_kernel_1 is both equal and unequal to Dot_ExpandDims_1.
(That is the output of the test I added, with some nuisance duplicate and reflexive constraints remove.)

**Comment by lesters** [comment]
Well, that's an effect of allowing !=, I did see the same during initial development. There was a comment left in there:

```
     * This requires having constraints that result in an absolute ordering:
     * equals, lesserThan and greaterThan do that, but adding notEquals does
     * not typically result in a guaranteed ordering. If that is needed, the
     * algorithm below needs to be adapted with a backtracking (tree) search
     * to find solutions.
```

This requires you to have some notion of the preferred order when renaming the dimensions for an operation. Is != really required here?

**Comment by bratseth** [comment]
I don't think it is a consequence of adding != because the constrains here have no solution. Yes we need != whenever we add a tensor (constant or argument) with multiple dimensions because otherwise there's nothing stopping the algorithm from choosing a solution where those two dimensions are equal, which is precisely what is does in this case.

The MatMul adds one more constraint than can be handled, we need to inject another dimension.

**Comment by bratseth** [comment]
Added a proof of concept which injects a rename operation which make this work.
Missing:
- Identify the right place to inject the rename
- Cleanup

I'll leave this for a while now as I really need to do some other things, in the meantime let me know if you do something on this.

**Comment by bratseth** [comment]
I believe this is ready to merge now. Ideally we should also do the tree search, consider more than one simultaneous rename if one is not sufficient, and consider non-root operations as rename targets.

I'm on vacation next week and may not be able to do work.

File: model-integration/src/main/java/ai/vespa/rankingexpression/importer/operations/IntermediateOperation.java [hunk 1]
```diff
@@ -99,6 +102,20 @@
     /** Add dimension name constraints for this operation */
     public void addDimensionNameConstraints(DimensionRenamer renamer) { }
 
+    /** Conveinence method to adds dimensions and constraints of the given tensor type */
```
Comment (lesters) [inline model-integration/src/main/java/ai/vespa/rankingexpression/importer/operations/IntermediateOperation.java]: * "Convenience method to add"

**Review by lesters — APPROVED** [review]
Ok, LGTM. Let's try this out... 👍

**Comment by bratseth** [comment]
We should also prefer to rename smaller tensors.

**Pull request merged**